freqtrade/rpc/emc.py

```diff
@@ -0,0 +1,234 @@
+    other FreqTrade bot's
+    """
+
+    def __init__(
+        self,
+        rpc: RPC,
+        config: Dict[str, Any],
+    ):
+        self._rpc = rpc
+        self._config = config
+
+        self._running = False
+        self._thread = None
+        self._loop = None
+        self._main_task = None
+        self._sub_tasks = None
+
+        self._emc_config = self._config.get('external_message_consumer', {})
+
+        self.enabled = self._emc_config.get('enabled', False)
+        self.producers = self._emc_config.get('producers', [])
+
+        if self.enabled and len(self.producers) < 1:
+            raise ValueError("You must specify at least 1 Producer to connect to.")
+
+        self.reply_timeout = self._emc_config.get('reply_timeout', 10)
+        self.ping_timeout = self._emc_config.get('ping_timeout', 2)
+        self.sleep_time = self._emc_config.get('sleep_time', 5)
+
+        # Setting these explicitly as they probably shouldn't be changed by a user
+        # Unless we somehow integrate this with the strategy to allow creating
+        # callbacks for the messages
+        self.topics = [RPCMessageType.WHITELIST, RPCMessageType.ANALYZED_DF]
+
+        self.start()
+
+    def start(self):
+        """
+        Start the main internal loop in another thread to run coroutines
+        """
+        self._loop = asyncio.new_event_loop()
+
+        if not self._thread:
+            logger.info("Starting ExternalMessageConsumer")
+
+            self._thread = Thread(target=self._loop.run_forever)
+            self._thread.start()
+            self._running = True
+        else:
+            raise RuntimeError("A loop is already running")
+
+        self._main_task = asyncio.run_coroutine_threadsafe(self._main(), loop=self._loop)
+
+    def shutdown(self):
+        """
+        Shutdown the loop, thread, and tasks
+        """
+        if self._thread and self._loop:
+            logger.info("Stopping ExternalMessageConsumer")
+
+            if self._sub_tasks:
+                # Cancel sub tasks
+                for task in self._sub_tasks:
+                    task.cancel()
+
+            if self._main_task:
+                # Cancel the main task
+                self._main_task.cancel()
+
+            self._thread.join()
+
+    async def _main(self):
+        """
+        The main task coroutine
+        """
+        rpc_lock = asyncio.Lock()
+
+        try:
+            # Create a connection to each producer
+            self._sub_tasks = [
+                self._loop.create_task(self._handle_producer_connection(producer, rpc_lock))
+                for producer in self.producers
+            ]
+
+            await asyncio.gather(*self._sub_tasks)
+        except asyncio.CancelledError:
+            pass
+        finally:
+            # Stop the loop once we are done
+            self._loop.stop()
+
+    async def _handle_producer_connection(self, producer, lock):
```

Comment:
we'd want to have typehints - that'll make our code safer by having mypy double-check us

Reply:
:+1: